WIP: Results TEP

wlynch opened the pull request:
local PR for early comments

imjasonh commented on teps/0017-results-api.md:
```diff
@@ -0,0 +1,711 @@
+---
+title: Tekton Results API
+authors:
+  - "@ImJasonH"
```
FWIW you can feel free to mark me as a reviewer and yourself as the sole author.

wlynch replied:
Half of this doc I straight copied from your original project proposal 😅

imjasonh commented on teps/0017-results-api.md:
```diff
@@ -0,0 +1,711 @@
+
+### Proposed Spec
+
+We propose the following API, heavily influenced by API recommendations defined
```
I'd like to propose an alternative that's been kicking around in my head, and get your feedback:

Instead, the Results API is a superset of the API-spec-conformant TaskRun/PipelineRun API, with the addition of a non-K8s-standard `filter` field on the API endpoint to list TaskRuns and PipelineRuns (i.e., `GET /apis/tekton.dev/v1beta1/{parent=*}/taskruns`). All other API methods work exactly the same, no changes, except responses are served from mysql instead of etcd.

This makes the K8s API server a slightly degenerate implementation of the Results API, since it serves GetTaskRuns/ListTaskRuns/etc. but just doesn't support `filter`.

The benefit is that taken holistically, a client doesn't have to think about "TaskRuns", "PipelineRuns", and "Results" as separate resources/endpoints, they can just think about the first two and maybe get better indexing/filtering depending on the backend.

This proposal isn't even really at odds with yours, we could probably do both and have this Tekton-API-compliant-plus-`filter` version be a facade for the `Resuls`-focused implementation.

WDYT?

wlynch replied:
I do see value in the facade based approach as an additional layer on top of the base Results API for Pipeline API compatibility - this effectively just functions as a selective view over results for a single type. This would likely make it easier for existing tools (tkn, dashboard) to adopt the results server without making significant changes.

However, supporting other Create/Delete/Update operations seems like something that should be handled by something other than the result server, so that the server can just focus on storing and serving result data. Another idea to go about doing this with a similar effect might be to have a meta Tekton control plane that knows to either call the Results API with a given query filter as a view for TaskRuns/PipelineRuns, else forward the request to the underlying k8s API server for mutating operations for all others.

> The benefit is that taken holistically, a client doesn't have to think about "TaskRuns", "PipelineRuns", and "Results" as separate resources/endpoints

I think this is part of the reason why we want a separate Results resource - when I want to see the things that ran in my cluster, I generally don't make a large distinction between a single task run or multiple task runs that were executed as part of a pipeline - I generally want to see a dashboard where each item is an event (e.g. trigger, cli command, etc) that caused an execution. The result type is meant to be this abstraction, so that we can group these for users as well as provide a mechanism to include other data that is related to Task/PipelineRuns, but doesn't have a good home at the moment (e.g. post-run notification statuses, meta-configs / DSLs, trigger information).

Looking ahead towards Trigger events, some results may not include executions at all, which we can't model with the TaskRuns and PipelineRuns API.

imjasonh replied:
Yeah I think you've convinced me that we need Results as a concept, and this is great in going in that direction.

Do you think you want to mention the facade idea in the TEP? Or maybe wait and leave it as a topic for discussion in an issue in the new repo?

wlynch replied:
Let's leave it off for now - it's an additive change and I don't believe it changes what we want to do in this proposal.

imjasonh commented on teps/0017-results-api.md:
````diff
@@ -0,0 +1,711 @@
+}
+```
+
+- API Server implementations must support all methods and fields within the
````
My advice would be to leave specifics of the API, and especially any concept of "conformance", outside the TEP.

The TEP is trying to convey "we should do X" with just enough detail about X to convince people  you have an idea how you'd do it (e.g., outline of a `Results` type and "there will be a CRUD API").

Diving into this much detail runs the risk of bikeshedding that would make the TEP take longer to land, which is a shame since I think everyone agrees we want this, and probably pretty urgently.

Once the TEP lands you can create a repo and make whatever changes you want with whatever other contributors/stakeholders you find.

wlynch replied:
Done. I feel like I might have gone a little too far in the other direction, so let me know what areas I should add more detail for.

imjasonh commented on teps/0017-results-api.md:
```diff
@@ -0,0 +1,711 @@
+-->
+
+We be consistent with the Pipelines
+[API compatibility policy](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md).
```
I'd state clearly this means that while Results is in alpha you can make any breaking changes you want.

wlynch replied:
Done.

imjasonh reviewed (commented):
🎉 I love it, ship it yesterday. 🎉

wlynch committed:
TEP-0021 Result API: Record Subresources

- Renames Event -> Record based on feedback (still has the same intent)
- Breaks Records out as a subresource of Results to allow for pagination
  and finer grained access.

wlynch committed:
TEP-0021 Results API: Mark TEP as implementable.
